core/src/main/java/io/questdb/griffin/engine/groupby/SampleByInterpolateRecordCursorFactory.java

```diff
@@ -500,24 +522,24 @@ protected RecordCursor initFunctionsAndCursor(
 
     private void interpolateBoundaryRange(long x1, long x2, Record record) {
         //interpolating boundary
-        for (int i = 0, m = groupByDistanceFunctions.size(); i < m; i++) {
-            GroupByFunction function = groupByDistanceFunctions.getQuick(i);
-            MapValue value1 = findDataMapValue2(record, x1);
-            MapValue value2 = findDataMapValue3(record, x2);
-            interpolatorBoundaryDistanceFunctions.getQuick(i).interpolateBoundaryAndStore(function, sampler.nextTimestamp(x1), value1, value2, true);
-            interpolatorBoundaryDistanceFunctions.getQuick(i).interpolateBoundaryAndStore(function, x2, value1, value2, false);
+        for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
+            GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
+            MapValue startValue = findDataMapValue2(record, x1);
+            MapValue endValue = findDataMapValue3(record, x2);
+            interpolatorBoundaryTwoPointFunctions.getQuick(i).interpolateBoundaryAndStore(function, sampler.nextTimestamp(x1), startValue, endValue, true);
+            interpolatorBoundaryTwoPointFunctions.getQuick(i).interpolateBoundaryAndStore(function, x2, startValue, endValue, false);
         }
     }
 
     private void interpolateDistanceRange(long x1, long x2, long lo, long hi, Record record) {
         for (long x = lo; x < hi; x = sampler.nextTimestamp(x)) {
-            final MapValue xValue = findDataMapValue(record, x);
-            assert xValue != null && xValue.getByte(0) == 1;
-            for (int i = 0, m = groupByDistanceFunctions.size(); i < m; i++) {
-                GroupByFunction function = groupByDistanceFunctions.getQuick(i);
-                MapValue value1 = findDataMapValue2(record, x1);
-                MapValue value2 = findDataMapValue3(record, x2);
-                interpolatorGapDistanceFunctions.getQuick(i).interpolateGapAndStore(function, xValue, sampler.getBucketSize(), value1, value2);
+            final MapValue result = findDataMapValue(record, x);
+            assert result != null && result.getByte(0) == 1;
+            for (int i = 0; i < groupByTwoPointFunctionCount; i++) {
+                GroupByFunction function = groupByTwoPointFunctions.getQuick(i);
+                MapValue startValue = findDataMapValue2(record, x1);
```

Comment:
This is excessive hash map scan. In some cases map values are already known, we shouldn't be looking up things that we don't have to

Reply:
done